From 2f5f4abc60bc92599fec0965d583bdbbbef94199 Mon Sep 17 00:00:00 2001
From: Peng Wu <wupeng58@huawei.com>
Date: Tue, 19 Apr 2022 06:15:43 +0000
Subject: [PATCH 3533/4585] [Huawei] mm: Add kernel param for memory reliable

Offering: HULK
hulk inclusion
category: feature
bugzilla: 187270

------------------------------------------------

Add kernel param reliable_debug in reparation for control memory reliable
features.

Signed-off-by: Peng Wu <wupeng58@huawei.com>
---
 .../admin-guide/kernel-parameters.txt         |  5 ++++
 mm/mem_reliable.c                             | 24 +++++++++++++++++++
 2 files changed, 29 insertions(+)

diff --git a/Documentation/admin-guide/kernel-parameters.txt b/Documentation/admin-guide/kernel-parameters.txt
index 0c21a75649d6..e45699e8561d 100644
--- a/Documentation/admin-guide/kernel-parameters.txt
+++ b/Documentation/admin-guide/kernel-parameters.txt
@@ -4780,6 +4780,11 @@
 			[KNL, SMP] Set scheduler's default relax_domain_level.
 			See Documentation/admin-guide/cgroup-v1/cpusets.rst.
 
+	reliable_debug=	[ARM64]
+			Format: []
+			Only works with CONFIG_MEMORY_RELIABLE and
+			"kernelcore=reliable" is configured.
+
 	reserve=	[KNL,BUGS] Force kernel to ignore I/O ports or memory
 			Format: <base1>,<size1>[,<base2>,<size2>,...]
 			Reserve I/O ports or memory so the kernel won't use
diff --git a/mm/mem_reliable.c b/mm/mem_reliable.c
index e0c410cddf7c..1ca67ede36cf 100644
--- a/mm/mem_reliable.c
+++ b/mm/mem_reliable.c
@@ -117,3 +117,27 @@ void reliable_report_meminfo(struct seq_file *m)
 	show_val_kb(m, "ReliableUsed:     ", used_reliable_pages());
 	show_val_kb(m, "ReliableBuddyMem: ", free_reliable_pages());
 }
+
+static int __init setup_reliable_debug(char *str)
+{
+	if (*str++ != '=' || !*str)
+		/*
+		 * No options specified.
+		 */
+		goto out;
+
+	/*
+	 * Determine which debug features should be switched on
+	 */
+	for (; *str && *str != ','; str++) {
+		switch (*str) {
+		default:
+			pr_err("reliable_debug option '%c' unknown. skipped\n",
+			       *str);
+		}
+	}
+
+out:
+	return 1;
+}
+__setup("reliable_debug", setup_reliable_debug);
-- 
2.17.1

